osquery/core/watcher.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+ *  Copyright (c) 2014, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <cstring>
+#include <sstream>
+
+#include <sys/wait.h>
+#include <signal.h>
+
+#include <osquery/core.h>
+#include <osquery/logger.h>
+#include <osquery/sql.h>
+
+#include "osquery/core/watcher.h"
+
+namespace osquery {
+
+#define WORKER_MEMORY_LIMIT (20 * 1024 * 1024)
+#define WORKER_UTILIZATION_LIMIT 60
+#define WORKER_RESPAWN_LIMIT 20
+#define WORKER_RESPAWN_DELAY 5
+#define WORKER_LATENCY_LIMIT 5
+
+bool Watcher::ok() {
+  ::sleep(1);
+  return (worker_ >= 0);
+}
+
+bool Watcher::watch() {
+  int status;
+  pid_t result = waitpid(worker_, &status, WNOHANG);
+  if (worker_ == 0 || result == worker_) {
+    // Worker does not exist or never existed.
+    return false;
+  } else if (result == 0) {
+    // If the inspect finds problems it will stop/restart the worker.
+    if (!isWorkerSane()) {
+      stopWorker();
+      return false;
+    }
+  }
+  return true;
+}
+
+void Watcher::stopWorker() {
+  kill(worker_, SIGKILL);
+  worker_ = 0;
+  // Clean up the defunct (zombie) process.
+  waitpid(-1, 0, 0);
```

Comment:
This is a blocking call, this could potentially lead to a lock in the watcher and should be called in a thread. In practice, I'm not sure if we could trigger a locking condition. Maybe if you post-exec changed the process' permissions?